cranelift/codegen/src/isa/x64/inst/emit.rs

```diff
@@ -4089,3 +4139,156 @@ pub(crate) fn emit(
 
     state.clear_post_insn();
 }
+
+/// Emit the common sequence used for both direct and indirect tail calls:
+///
+/// * Copy the new frame's stack arguments over the top of our current frame.
+///
+/// * Restore the old frame pointer.
+///
+/// * Initialize the tail callee's stack pointer (simultaneously deallocating
+///   the temporary stack space we allocated when creating the new frame's stack
+///   arguments).
+///
+/// * Move the return address into its stack slot.
+fn emit_return_call_common_sequence(
+    allocs: &mut AllocationConsumer<'_>,
+    sink: &mut MachBuffer<Inst>,
+    info: &EmitInfo,
+    state: &mut EmitState,
+    new_stack_arg_size: u32,
+    old_stack_arg_size: u32,
+    ret_addr: Gpr,
+    fp: Gpr,
+    tmp: WritableGpr,
+    uses: &CallArgList,
+) {
+    assert!(
+        info.flags.preserve_frame_pointers(),
+        "frame pointers aren't fundamentally required for tail calls, \
+                 but the current implementation relies on them being present"
+    );
+
+    for u in uses {
+        let _ = allocs.next(u.vreg);
+    }
+
+    let ret_addr = allocs.next(*ret_addr);
+    let ret_addr = Gpr::new(ret_addr).unwrap();
+
+    let fp = allocs.next(*fp);
+
+    let tmp = allocs.next(tmp.to_reg().to_reg());
+    let tmp = Gpr::new(tmp).unwrap();
+    let tmp_w = WritableGpr::from_reg(tmp);
+
+    // Copy the new frame (which is `frame_size` bytes above the SP)
+    // onto our current frame, using only volatile, non-argument
+    // registers.
+    //
+    //
+    // The current stack layout is the following:
+    //
+    //            | ...                 |
+    //            +---------------------+
+    //            | ...                 |
+    //            | stack arguments     |
+    //            | ...                 |
+    //    current | return address      |
+    //    frame   | old FP              | <-- FP
+    //            | ...                 |
+    //            | old stack slots     |
+    //            | ...                 |
+    //            +---------------------+
+    //            | ...                 |
+    //    new     | new stack arguments |
+    //    frame   | ...                 | <-- SP
+    //            +---------------------+
+    //
+    // We need to pop FP, copy the new stack arguments over the old
```

Comment:
possibly just a semantic quibble but "pop FP" confused me here: to pop means to adjust SP, but here we're restoring FP's old value without adjusting the stack at all. Could we write it that way ("restore FP") instead?

Reply:
Done.